---
title: "RepViz-vignette"
author: "Thomas Faux, Kalle RytkÃ¶nen, Asta Laiho, Laura L. Elo"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{RepViz}
  %\VignetteEncoding{UTF-8}
  %\usepackage[UTF-8]{inputenc}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

DNA sequencing has become an essential part of biomedicine and biology. Commonly, the data are analyzed with automatized pipelines. However, a visual inspection by a researcher is often useful both at the level of basic quality control and as a confirmation of the analysis results. Efficient visualization often has an essential role also in guiding the analysis design and interpretation of the results.
In an effort to reduce the amount of data in the visualization output and to optimize the information for the user, we have developed a replicate driven tool, RepViz, which enables to view a genomic region in a simple and efficient way. RepViz allows simultaneous viewing of both intra- and intergroup variation in sequencing counts of the studied conditions, as well as their comparison to the output features (e.g. identified peaks) from user selected data analysis methods. The RepViz tool is primarily designed for chromatin data such as ChIP-seq and ATAC-seq, but can also be used with other sequencing data such as RNA-seq, or combinations of different types of genomic data.

## The input files

The tool takes two CSV files as an input. These CSV files are contains the paths to the different files needed for RepViz.One of the CSV file containes the file paths to the BAM files and organises them by group and the second CSV file gives the file paths of the BED files. You can see below an exemple of the BAM_input.csv containing three columns, the path to the BAM files, the group number and the group names.  

### The BAM input file:
```{r input BAM}
BAM_table <- read.table(system.file("extdata","BAM_input.csv", package = "RepViz"), sep = ";")
colnames(BAM_table) <- c("bam file","group","group name")
knitr::kable(BAM_table)
```

You can see below an exemple of the BED_input.csv containing two columns, the path to the BED files and the name that will appear in the legend. For the purpose of this exemple the file contains only one line, but it is meant to contain multiple lines.
### The BED input file
```{r input BED}
BED_table <- read.table(system.file("extdata","BED_input.csv", package = "RepViz"), sep = ";")
colnames(BED_table) <- c("bed file","Legend")
knitr::kable(BED_table)
```

Note that the corresponding BAI files needs to be present in the same folder with the BAM files 

```{r bam files, echo=FALSE}
list.files(system.file("extdata", package = "RepViz"),pattern = ".bam")
```

## The plotting function

Once the CSV input files are ready you need to declare the region that will be drawn. For this exemple we are using the region in the vicinity of the VPS29 gene. The library GenomicRanges will be used to create a GRanges object containing the region. The genome needs to be specified in order to have the genomic track matching your other tracks (BAM, BED). The genomes that are currently implemented are hg19, hg38 and mm10. The arguments avgTrack and geneTrack are logical and enable you to control whether the gene track and average track are plotted or not.

```{r GRanges, echo=FALSE,prompt=FALSE}
suppressMessages(library(GenomicRanges))
```
```{r plot,fig.height=10,fig.width=12}
region <- GRanges("chr12:110938000-110940000")
backup <- getwd()
setwd(system.file("extdata", package = "RepViz"))
RepViz::RepViz(region = region,
               genome = "hg19",
               BAM = "BAM_input.csv",
               BED = "BED_input.csv",
               avgTrack = T,
               geneTrack = T)



```

This line sets the working directory back in his original place
```{r backup}
setwd(backup)
```
