\documentclass[a4paper]{book}
\usepackage[times,inconsolata,hyper]{Rd}
\usepackage{makeidx}
\usepackage[utf8]{inputenc} % @SET ENCODING@
% \usepackage{graphicx} % @USE GRAPHICX@
\makeindex{}
\begin{document}
\chapter*{}
\begin{center}
{\textbf{\huge Package}}
\par\bigskip{\large \today}
\end{center}
\begin{description}
\raggedright{}
\inputencoding{utf8}
\item[Type]\AsIs{Package}
\item[Title]\AsIs{Replicate oriented Visualization of a genomic region}
\item[Version]\AsIs{0.1.12}
\item[Authors]\AsIs{Thomas Faux, Kalle RytkÃ¶nen, Asta Laiho, Laura L. Elo}
\item[Maintainer]\AsIs{The package maintainer }\email{faux.thomas1@gmail.com}\AsIs{}
\item[Description]\AsIs{RepView enables the view of a genomic region in a simple and efficient way.RepView allows simultaneous viewing of both intra- and intergroup variation in sequencing counts of the studied conditions, as well as their comparison to the output features (e.g. identified peaks) from user selected data analysis methods.The RepView tool is primarily designed for chromatin data such as ChIP-seq and ATAC-seq, but can also be used with other sequencing data such as RNA-seq, or combinations of different types of genomic data.}
\item[License]\AsIs{GPL-3}
\item[Encoding]\AsIs{UTF-8}
\item[LazyData]\AsIs{true}
\item[RoxygenNote]\AsIs{6.1.1}
\item[VignetteBuilder]\AsIs{knitr}
\item[Depends]\AsIs{R (>= 3.4.0),
GenomicRanges (>= 1.30.0),
rbamtools (>= 2.16.11),
IRanges (>= 2.14.0),
biomaRt (>= 2.36.0),
S4Vectors (>= 0.18.0)}
\item[Suggests]\AsIs{knitr,
testthat}
\end{description}
\Rdcontents{\R{} topics documented:}
\inputencoding{utf8}
\HeaderA{RepViz}{Plot a genomic region}{RepViz}
%
\begin{Description}\relax
Plot a genomic region
\end{Description}
%
\begin{Usage}
\begin{verbatim}
RepViz(region, genome, BAM = NULL, BED = NULL, avgTrack = TRUE,
  geneTrack = TRUE, max = NULL, verbose = TRUE)
\end{verbatim}
\end{Usage}
%
\begin{Arguments}
\begin{ldescription}
\item[\code{region}] a GRange object with chr, start, end

\item[\code{genome}] a character vector "hg19","hg38" or "mm10"

\item[\code{BAM}] a path to the BAM related csv input file

\item[\code{BED}] a path to the BED related csv input file

\item[\code{avgTrack}] a logical indicating if the average track should be present or not

\item[\code{geneTrack}] a logical indicating if the gene track should be present or not

\item[\code{max}] a vector of number containing the maximum of each BAM track

\item[\code{verbose}] prompt the progress of the plotting
\end{ldescription}
\end{Arguments}
%
\begin{Examples}
\begin{ExampleCode}
backup <- getwd()
region <- GRanges("chr12:110938000-110940000")
setwd(system.file("extdata", package = "RepViz"))
RepViz::RepViz(region = region,
               genome = "hg19",
               BAM = "BAM_input.csv",
               BED = "BED_input.csv",
               avgTrack = TRUE,
               geneTrack = TRUE)
setwd(backup)

\end{ExampleCode}
\end{Examples}
\printindex{}
\end{document}
